### PR TITLE
[SD-LIE] Multiple fractures

### PR DESCRIPTION
**Pull request by norihiro-w**
This PR introduces a support of multiple non-intersecting fractures in SD-LIE process. Major changes are
- extend `getFractureMatrixDataInMesh()` to looking for multiple fracture groups having distinct material IDs
- extend the local assembler for matrix near fractures. Now the displacement function is given as `u(x) =  N(x) {u} + sum_i [phi_i(x) N(x) {g_i}]` where `phi_i(x)` is the levelest function, `g_i` is  the nodal enrichment values for `i` th fracture.
- extend the post processing tool

### CODE REVIEW

**Comment by norihiro-w** [comment]
@endJunction I got MSVC error C2797 ('SmallDeformationProcessData::_fracture_property': list initialization inside member initializer list or non-static data member initializer is not implemented
) for the following code.  Do you have any idea how to achieve move of a vector of unique pointers?

```
struct SmallDeformationProcessData
{
    SmallDeformationProcessData(SmallDeformationProcessData&& other)
        : _fracture_property{std::move(other._fracture_property)} // error C2797
    {
    }

    std::vector<std::unique_ptr<FractureProperty>> _fracture_property;
}
```

**Comment by chleh** [comment]
@norihiro-w, maybe you have to use round brackets, not curly braces:
`: _fracture_property(std::move(other._fracture_property))`
instead of
`: _fracture_property{std::move(other._fracture_property)}`.

**Comment by norihiro-w** [comment]
@chleh it worked! thanks!

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 1]
```diff
@@ -115,21 +118,60 @@ assembleWithJacobian(
...
+    //--------------------------------------------------------------------------------------
+    // prepare sub vectors, matrices for regular displacement (u) and displacement jumps (g)
+    //
+    // example with two fractures:
+    //     |b(u)|
+    // b = |b(g1)|
+    //     |b(g2)|
+    //
+    //     |J(u,u)  J(u,g1)  J(u,g2) |
+    // J = |J(g1,u) J(g1,g1) J(g2,g2)|
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: g1,g2

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp [hunk 2]
```diff
@@ -62,73 +63,106 @@ class IsCrackTip
...
-         vec_matrix_elements.size(), vec_fracture_elements.size());
+         vec_matrix_elements.size(), all_fracture_elements.size());
+
+    // get fracture material IDs
+    auto opt_material_ids(mesh.getProperties().getPropertyVector<int>("MaterialIDs"));
+    if (!opt_material_ids)
+        OGS_FATAL("MaterialIDs propery vector not found in a mesh");
+    for (MeshLib::Element* e : all_fracture_elements)
+        vec_fracture_mat_IDs.push_back((*opt_material_ids)[e->getID()]);
+    BaseLib::makeVectorUnique(vec_fracture_mat_IDs);
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: Minor: acutally it looks like `std::set` would be a better type for `vec_fracture_mat_IDs`.

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp [hunk 3]
```diff
@@ -62,73 +63,106 @@ class IsCrackTip
...
+        for (MeshLib::Element* e : vec_fracture_elements[frac_id])
         {
-            if (isCrackTip(*e->getNode(i)))
-                continue;
-            vec_fracture_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
+            for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+            {
+                if (isCrackTip(*e->getNode(i)))
+                    continue;
+                vec_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: Just wondering: Can't the nodes be `const`? (Then we probably need `std::vector<std::vector<MeshLib::Node const*>>& vec_fracture_nodes`, too)

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp [hunk 4]
```diff
@@ -62,73 +63,106 @@ class IsCrackTip
...
-        for (unsigned i=0; i<e->getNumberOfBaseNodes(); i++)
+        std::vector<MeshLib::Element*> vec_ele;
+        // first, collect matrix elements
+        for (MeshLib::Element*e : fracture_elements)
         {
-            MeshLib::Node const* node = e->getNode(i);
-            if (isCrackTip(*node))
-                continue;
-            for (unsigned j=0; j<node->getNumberOfElements(); j++)
+            for (unsigned i=0; i<e->getNumberOfBaseNodes(); i++)
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: Maybe add a comment that it is sufficient to iterate over base nodes, because they are already connected to all neighbours?

File: ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp [hunk 5]
```diff
@@ -52,43 +52,56 @@ PostProcessTool::PostProcessTool(
...
     {
-        auto duplicated_node = new MeshLib::Node(org_node->getCoords(), new_nodes.size());
-        new_nodes.push_back(duplicated_node);
-        _map_dup_newNodeIDs[org_node->getID()] = duplicated_node->getID();
+        for (auto const* org_node : vec_fracture_nodes)
+        {
+            auto duplicated_node = new MeshLib::Node(org_node->getCoords(), new_nodes.size());
+            new_nodes.push_back(duplicated_node);
+            assert(_map_dup_newNodeIDs.count(org_node->getID())==0);
+            _map_dup_newNodeIDs[org_node->getID()] = duplicated_node->getID();
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp]: Minor: I'd do the check like this (if `_map_dup_newNodeIDs` is a map):
```{.cpp}
if (! _map_dup_newNodeIDs.emplace(org_node->getID(), duplicated_node->getID()).second)
    OGS_FATAL("error ...");
```

File: ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp [hunk 6]
```diff
@@ -52,43 +52,56 @@ PostProcessTool::PostProcessTool(
...
-        auto const eid = org_e->getID();
-
-        // keep original if the element has levelset=0
-        if ((*prop_levelset)[eid] == 0)
-            continue;
-
-        // replace fracture nodes with duplicated ones
-        MeshLib::Element* e = new_eles[eid];
-        for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+        auto const& vec_fracutre_matrix_elements = vec_vec_fracutre_matrix_elements[fracture_id];
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp]: typo `fracutre`

File: ProcessLib/SmallDeformationWithLIE/CreateSmallDeformationProcess.cpp [hunk 7]
```diff
@@ -136,16 +137,26 @@ createSmallDeformationProcess(
...
-    auto fracture_properties_config = config.getConfigSubtree("fracture_properties");
-    auto &para_b0 = ProcessLib::findParameter<double>(fracture_properties_config, "initial_aperture", parameters, 1);
-    std::unique_ptr<FractureProperty> frac_prop(new FractureProperty());
-    frac_prop->mat_id = fracture_properties_config.getConfigParameter<int>("material_id");
-    frac_prop->aperture0 = &para_b0;
+    for (auto fracture_properties_config : config.getConfigSubtreeList("fracture_properties"))
+    {
+        auto& para_b0 = ProcessLib::findParameter<double>(
+            fracture_properties_config, "initial_aperture", parameters, 1);
+        auto frac_prop(new FractureProperty());
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/CreateSmallDeformationProcess.cpp]: Using std::unique_ptr directly would be better.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 8]
```diff
@@ -115,21 +118,60 @@ assembleWithJacobian(
...
+    //--------------------------------------------------------------------------------------
+    auto local_b_u = local_b.segment<N_DOF_PER_VAR>(0);
+    std::vector<BlockVectorType> vec_local_b_g;
+    for (unsigned i=0; i<n_fractures; i++)
+        vec_local_b_g.push_back(local_b.segment<N_DOF_PER_VAR>(N_DOF_PER_VAR*(i+1)));
+
+    auto local_J_uu = local_J.block<N_DOF_PER_VAR, N_DOF_PER_VAR>(0, 0);
+    std::vector<BlockMatrixType> vec_local_J_ug;
+    std::vector<BlockMatrixType> vec_local_J_gu;
+    std::vector<std::vector<BlockMatrixType>> vec_local_J_gg(n_fractures);
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Maybe reserve sizes for the vectors?

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 9]
```diff
@@ -163,30 +209,36 @@ assembleWithJacobian(
...
-        // r_ru = B^T Stress = B^T C B (u + phi*[u])
-        localRhs_ru.noalias() -= B.transpose() * sigma * detJ * wp.getWeight() * integralMeasure;
-
-        // r_[u] = (phi*B)^T Stress = (phi*B)^T C B (u + phi*[u])
-        localRhs_du.noalias() -= levelsets * B.transpose() * sigma * detJ * wp.getWeight() * integralMeasure;
+        // r_u = B^T * Sigma = B^T * C * B * (u+phi*[u])
+        // r_[u] = (phi*B)^T * Sigma = (phi*B)^T * C * B * (u+phi*[u])
+        local_b_u.noalias() -= B.transpose() * sigma * ip_factor;
+        for (unsigned i=0; i<n_fractures; i++)
+            vec_local_b_g[i].noalias() -= levelsets[i] * B.transpose() * sigma * ip_factor;
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Is it really better to store the Eigen::Blocks in a std:vector instead of building them on demand? E.g., `vec_local_b_g` seems to be used only here.

File: ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp [hunk 10]
```diff
@@ -57,25 +57,57 @@ SmallDeformationProcess<DisplacementDim>::SmallDeformationProcess(
...
-                        *_process_data._fracture_property.get());
+    if (_vec_fracture_mat_IDs.size() != _process_data._vec_fracture_property.size())
+        OGS_FATAL("The number of the given fracture properties (%d) are not consistent"
+                  " with the number of fracture groups in a mesh (%d).",
+                  _process_data._vec_fracture_property.size(),
+                  _vec_fracture_mat_IDs.size());
+
+    // create a map from a material ID to a fracture ID
+    auto max_frac_mat_id = std::max_element(_vec_fracture_mat_IDs.begin(),
+                                            _vec_fracture_mat_IDs.end());
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp]: Currently fractures are identified with some material IDs. Does that also work, if somebody already has multiple materials in the matrix (i.e., non-fracture part)?

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 11]
```diff
@@ -163,30 +209,36 @@ assembleWithJacobian(
...
-        // J_[u][u] += (levelset B)^T * C * (levelset B)
-        localA_gg.noalias() += levelsets * B.transpose() * C * levelsets * B * detJ * wp.getWeight() * integralMeasure;
+        local_J_uu.noalias() += B.transpose() * C * B * ip_factor;
+
+        for (unsigned i=0; i<n_fractures; i++)
+        {
+            // J_u[u] += B^T * C * (levelset * B)
+            vec_local_J_ug[i].noalias() += B.transpose() * C *(levelsets[i] * B) * ip_factor;
+
+            // J_[u]u += (levelset * B)^T * C * B
```
Comment (wenqing) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: According to the following line
``(levelset * B)^T`` should be ``(levelset * B^T)``

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 12]
```diff
@@ -163,30 +209,36 @@ assembleWithJacobian(
...
-
-        // J_u[u] += B^T * C * B * levelset
-        localA_ug.noalias() += B.transpose() * C * levelsets * B * detJ * wp.getWeight() * integralMeasure;
-
-        // J_[u]u += (levelset B)^T * C * B
-        localA_gu.noalias() += levelsets * B.transpose() * C * B * detJ * wp.getWeight() * integralMeasure;
-
-        // J_[u][u] += (levelset B)^T * C * (levelset B)
-        localA_gg.noalias() += levelsets * B.transpose() * C * levelsets * B * detJ * wp.getWeight() * integralMeasure;
+        local_J_uu.noalias() += B.transpose() * C * B * ip_factor;
```
Comment (wenqing) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: ``B^T * C`` is performed several times in the following two loops. If  n_fractures is a large number, a computational optimization might be needed here after.

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp (hunk 2 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: could be, but I would not include optimisation in this PR.

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp (hunk 3 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: no, because the vector will be passed to MeshSubset which accepts `vector<Node*> const*`.

File: ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp (hunk 6 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/PostUtils.cpp]: again...  sorry

File: ProcessLib/SmallDeformationWithLIE/CreateSmallDeformationProcess.cpp (hunk 7 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/CreateSmallDeformationProcess.cpp]: i don't see a benefit. keep it as it is

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 8 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: optimization will not be included in this PR.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 9 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: only here, but is called for each integration point.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 12 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: yes you are right. but I will not include the optimisation at this stage. Because I'll not maintain the LIE stuff and someone has to take over this, I think readability is more important.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 11 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Correct expression is `(levelset * B)^T`. Because the levelset I use is a scalar, we can code it as below.

File: ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp (hunk 10 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp]: yes, it works, because the code accesses `_process_data._map_materialID_to_fractureID` only with material IDs of fractures.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 8 above)
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Since this code is executed in every local assemble, I think it would be very good to do the reservation.
It's three lines of code. Please do it. We do it in many other places, too. It stresses good style.

File: ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp (hunk 10 above)
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/SmallDeformationProcess.cpp]: Thanks.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 8 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: as i said. i'll not include optimisation. if optimisation is necessary, one should first profile the code. if this part significantly slows down local assembly, one can fix it easily.

File: ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h (hunk 11 above)
Comment (wenqing) [inline ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: I see.
